### ISSUE TITLE
Implement Table Macro

### ISSUE THREAD

**Issue by hmb1** [duckdb/duckdb#2735]
Oracle and PostgreSQL both differentiate between scalar and table macros, however DuckDB currently only supports a scalar (or single column) macro. Looking at the DuckDB implementation of ‘CREATE MACRO’ in macro_function.cpp, it should be possible to add a field to indicate table/scalar macro. 

@Mytherin -- could you please let me know if you have any suggestions for its implementation?

The grammar I envisage is:
CREATE [{TABLE|SCALAR}] MACRO

and 'CREATE MACRO' will default to SCALAR so that nothing breaks.

Thanks
....Henry

**Comment by Mytherin** [comment]
That is a great idea and something we have been wanting to implement!

> @Mytherin -- could you please let me know if you have any suggestions for its implementation?

The most work here will be in the binding of the macro function, but many of the components of the current macro infrastructure could be re-used there. Since the current macro functions also support placing arguments inside subqueries (e.g. `CREATE MACRO f(x) AS (SELECT x+1);`), in principle most of the components are already there.

I would start by extending the parser, and having a look at the existing macro binding implementation (`src/planner/binder/expression/bind_macro_expression.cpp`). Many of those components should be re-usable for this purpose.

**Comment by hmb1** [comment]
@Mytherin
Thanks very much for the road map - just finding my feet.  
Is there a function/method in  duckdb/third_party/libpg_query   that pretty prints out the parse tree ?
Thanks

**Comment by Mytherin** [comment]
I don't think there is such a function, but I might be mistaken. I have never used it in any case.

**Comment by hmb1** [comment]
OK Many thanks

**Comment by hmb1** [comment]
@Mytherin 

Have  modified  create_function.y   so that it can parse the select macro e.g

**CREATE MACRO  TABLE  select_b(x) AS SELECT 1+x, 2+x;
CREATE MACRO TABLE select_filterName(_name := 'Henry') AS SELECT * FROM table WHERE name=_name;**

Have added following field to   CreateMacroInfo()
//! The SelectStatement of the macro
unique_ptr<SelectStatement> query;

What I don't get is the context where this select  macro should be  called ?
In a statement context  and not an expression context like regular macros and aggregates ?

For example 
**create macro ms(a) as  (select * from mytable);**
Error: Binder Error: Subquery returns 2 columns - expected 1
**create view select_filterName(_name := 'Joe')**

Perhaps its better to have a new class  "CreateMacroSelectInfo"  rather than appending to  "CreateMacroInfo"


I think I can bind the query in Macro i - by doing parameter/column name replacements and then using bind_select_node.cpp. 
 

Many thanks for any input.

**Comment by hmb1** [comment]
@Mytherin 
HI there  
  I have added some new public members to MacroFunc
  But the problem is  that when I  Serialize /  DeSerialize  it breaks the existing MacroFuncCatalog entry.  
 Your guidance in greatly appreciated .

 (  My New MacroFunc is below)
  Have added the  following members

```
        // if true then we have a query_node and not a regular expression
	bool is_query;
	//! The main query node
	unique_ptr<QueryNode> query_node;
```  
 Not sure if  I am talking rubbish but I see two options:     
 Create a new Macro type Say MacroFuncQuery() or  breaking the existing MacroCatalog is OK ?
 Thanks


``
class MacroCatalogEntry;


class MacroFunction {
public:
	explicit MacroFunction(unique_ptr<ParsedExpression> expression);

	//! Check whether the supplied arguments are valid
	static string ValidateArguments(MacroCatalogEntry &macro_func, FunctionExpression &function_expr,
	                                vector<unique_ptr<ParsedExpression>> &positionals,
	                                unordered_map<string, unique_ptr<ParsedExpression>> &defaults);
	//! The macro expression
	unique_ptr<ParsedExpression> expression;
	//! The positional parameters
	vector<unique_ptr<ParsedExpression>> parameters;
	//! The default parameters and their associated values
	unordered_map<string, unique_ptr<ParsedExpression>> default_parameters;

	// if true then we have a query_node and not a regular expression
	bool is_query;
	//! The main query node
	unique_ptr<QueryNode> query_node;


public:
	unique_ptr<MacroFunction> Copy();


};

**Comment by Mytherin** [comment]
> Create a new Macro type Say MacroFuncQuery() or breaking the existing MacroCatalog is OK ?

Breaking the existing serialization is fine, you will just need to increment the storage version in `src/storage/storage_info.cpp` and re-generate the stored database test.

> bool is_query;

Is there any reason for this boolean flag? I assume either `expression` or `query_node` will be `nullptr`, depending on which type of macro function it is, no?

**Comment by hmb1** [comment]
Thanks Mark, 
What about appending  query_node to the catalog entry  ?
Then their will be no break ?
Or Is this just an ugly hack ?

**Comment by Mytherin** [comment]
Hm, perhaps it is indeed better to create a separate catalog entry and function entry, as the functions do indeed have some differences. Perhaps the best solution here is to create a `BaseMacroFunction` that holds the common parts that the `ScalarMacroFunction` and `TableMacroFunction` can inherit from.

**Comment by hmb1** [comment]
Yeh,
 I am quite a long way down the road with my new MacroFunc ?

**Comment by hmb1** [comment]
@Mytherin 
Have got things working  working  can now define table macro as say

~~~
CREATE MACRO TABLE  xt(a,_name) as select id,name from test where (id >=a or name=_name);
CREATE MACRO TABLE  xt2(a,_name) as select id,name from test2 where (id >=a or name=_name);
CREATE MACRO TABLE m1(a,b)   as select a, b from maths;
~~~
and then use them as say
~~~
SELECT FUNCTION xt(10.'joe');  
SELECT FUNCTION xt2( 1,'henry');
SELECT FUNCTION m1(i*i, 5*j+10);
~~~
Can also use them in a SET query e.g  
~~~
SELECT FUNCTION xt(1,'fred') UNION SELECT FUNCTION x2(1,'mary')
~~~

The FUNCTION keyword in the select is used to indicate to the parser that this is a Table Macro and not a scalar Macro, Window or Aggregate function. e.g `SELECT abs(-10)`

The FUNCTION keyword  works  its way to my BindNodeMacro()  via the  new members   **PGFuncCall->is_select_macro**  and   **FunctionExpression->is_select_macro**.   The alternative to this is to add a new member to QueryNode to indicate that the select is a Macro.  Since coding I realise that either method will require a Catalog rewrite    ( of Serilaizer/Deserializer in  QueryNode or FunctionExpression).  Are mods to QueryNode  out of bounds? What are your thoughts ?
Many thanks

My code fork is at  ``https://github.com/hmb1/duckdb/commits/hmb1-select-macro-feature


...Henry

**Comment by hannesmuehleisen** [comment]
Can you just open a PR for us to review if you think its ready?

**Comment by hmb1** [comment]
@Mytherin 
Code is looking good but I still need to add a  TableMacroFunction  to the catalog 
 and I am running out of time. 
Any short cut ideas ?

**Comment by Mytherin** [comment]
You would need to create a new catalog entry, that is likely based on the existing macro catalog entry.

**Comment by hmb1** [comment]
@Mytherin 
Things are going well - nearly ready for a PR !!
I am now writing the Table Macros to the Table Function catalog and everything is working  except that Table Macros are only being stored temporarily in the DB -  I they don't  don't get written to the DB proper

I am guessing this is the behaviour you want for the regular Table Functions - and that  some temporary flag is set to true for this catalg but i cannot find it. ?
 thanks
...Henry

**Comment by Mytherin** [comment]
Regular table functions are not persisted indeed. I recommend looking at how this is done for the existing macro infrastructure, particularly grepping for `CatalogType::MACRO_ENTRY` in the code (e.g. commit_state.cpp, checkpoint_manager.cpp, wal_replay.cpp).

**Comment by hmb1** [comment]
Am having some issues with the storage version  
( in my branch I have not incremented storage version)
When I run the following test it fails 
**build/test/unittest test/sql/storage_version/storage_version.test_slow**
and I cannot read the db either
**build/duckdb test/sql/storage_version/storage_version.db**

However  when I recreate storage_version.db ( without incrementing the storage version ) using the python script generate_storage_version.py. The above 2 tests work. what is this indicating ?  

Thanks for any help

**Comment by Mytherin** [comment]
It indicates that the backwards compatible reading of the storage broke because of the changes in your PR. Likely this is because of changes made to the `Serialize` and `Deserialize` method of the macro functions.

**Comment by hmb1** [comment]
Many thanks mark -  
I'll take another look at them

**Comment by hmb1** [comment]
@Mytherin 
The issue is caused by my appending table macros in  `WriteSchema()`   
A new record of type unint32   and value table_macros.size is being appended. 
( along with ( if any)  the Table  Macro data )
 Can you think of a way of modifying my ReadSchema/writeSchema so that I don't have to increment the storage version ?

~~~
void CheckpointManager::WriteSchema(SchemaCatalogEntry &schema) 
...
...
         // finally write the macro's
	metadata_writer->Write<uint32_t>(table_macros.size());
	for (auto &macro : table_macros) {
		WriteTableMacro(*macro);
	}


void CheckpointManager::ReadSchema(ClientContext &context, MetaBlockReader &reader) 
...
...
        uint32_t table_macro_count = reader.Read<uint32_t>();
	for (uint32_t i = 0; i < table_macro_count; i++) {
		ReadTableMacro(context, reader);
	}
   ~~~

**Comment by Mytherin** [comment]
Hm, you are indeed correct, it is not possible to do so without breaking the storage since the FieldWriter/FieldReader classes are not used there yet. I will make a note of it and fix it later. Feel free to increment the storage version and add your code there. Thanks for the heads up!

**Comment by Mytherin** [comment]
This has been implemented in #3084
